### PR TITLE
fix unsubscribe

### PR DESCRIPTION
**Pull request by cskiraly**
Unsubscribe which was keeping the websocket open, leading to an accumulation of websocket connections. The correct was of unsubscribing from a subscription created by a websocket is to close the websocket.

@craig-mulligan, please check with agile-data
@Exulansis, please check with agile-ui

### CODE REVIEW

File: src/device/index.js [hunk 1]
```diff
@@ -174,7 +175,11 @@ const device = (base, wsBase) => {
     * });
     **/
     subscribe: (deviceId, componentId) => {
-      return Promise.resolve(new WS(`${wsBase}/${deviceId}/${componentId}/subscribe`));
+      var id = deviceId+componentId;
+      if (! websockets.hasOwnProperty(id)) {
```
Comment (craigmulligan) [inline src/device/index.js]: NIT: remove space
`! websockets.hasOwnProperty(id)` -> `!websockets.hasOwnProperty(id)`

File: src/device/index.js [hunk 2]
```diff
@@ -174,7 +175,11 @@ const device = (base, wsBase) => {
     * });
     **/
     subscribe: (deviceId, componentId) => {
-      return Promise.resolve(new WS(`${wsBase}/${deviceId}/${componentId}/subscribe`));
+      var id = deviceId+componentId;
```
Comment (craigmulligan) [inline src/device/index.js]: NIT: use `let` instead of `var`

File: src/device/index.js [hunk 3]
```diff
@@ -191,10 +196,13 @@ const device = (base, wsBase) => {
     *  console.log('Unsubscribed!');
     * });
     **/
-    unsubscribe: (deviceId, componentId) => axios({
-      method: 'DELETE',
-      url: `${base}/${deviceId}/${componentId}/subscribe`
-    })
+    unsubscribe: (deviceId, componentId) => {
+      var id = deviceId+componentId;
```
Comment (craigmulligan) [inline src/device/index.js]: NIT: use `Let`
NIT: space between concatination. `deviceId + componentId`

File: src/device/index.js [hunk 4]
```diff
@@ -191,10 +196,13 @@ const device = (base, wsBase) => {
...
+      if (websockets.hasOwnProperty(id)) {
+        websockets[id].close();
+        delete websockets[id];
```
Comment (craigmulligan) [inline src/device/index.js]: You should return a promise, as the api was previously `thenable`. You can do something like:

```
try {
    websockets[id].close();
    delete websockets[id];
     return Promise.resolve();
} catch (err) {
    return Promise.reject(err);
}
```

Comment (craigmulligan) [reply src/device/index.js]: Oh and do we no longer need to do a `DELETE` on `unsubscribe`? If not we should update the api docs.

Comment (craigmulligan) [reply src/device/index.js]: Oh also I see one line `190` the example doc is wrong. `agile.device.get` should be `agile.device.unsubscribe` Could you update with this PR please?

**Review by craigmulligan — COMMENTED** [review]
Couple of nits but otherwise looks good to me. I thought sending a `DELETE` to `/subscribe` the server would send the `CLOSE` event to the socket? Happy to merge if this is this is easier fix, but it instead be better to do on the server because custom (non-sdk) connections could leave dangling connections?

File: src/device/index.js (hunk 4 above)
Comment (cskiraly) [reply src/device/index.js]: subscribe and unsubscribe goes in pairs. You either use the HTTP version (and then a generic WS to get the messages), or the component specific WS. In this case, you subscribe by opening the WS, and unsubscribe by closing the WS. Note that there could be several WSs subscribed at the same time, and in the bacground at the protocol level we only unsubscribe when all are closed, and also eventual HTTP subscribes had been deleted.

@craig, can you implement the changes you've proposed? Thanks

**Comment by cskiraly** [comment]
It seems to me that the discovery status test fails because the test server is not updated. Could it be?

**Pull request merged**